From: Miroslav Benes <mbenes@suse.cz>
Date: Thu, 9 Apr 2015 17:28:45 +0200
Subject: kgr: send a fake signal to all blocking tasks
Patch-mainline: not yet, kgraft
References: fate#313296

kGraft consistency model is of LEAVE_KERNEL and SWITCH_THREAD. This
means that all tasks in the system have to be marked one by one as safe
to call a new patched function. Safe place is on the boundary between
kernel and userspace. The patching waits for all tasks to cross this
boundary and finishes the process afterwards.

The problem is that a task can block the finalization of patching
process for quite a long time, if not forever. The task could sleep
somewhere in the kernel or could be running in the userspace with no
prospect of entering the kernel and thus going through the safe place.

Luckily we can force the task to do that by sending it a fake signal,
that is a signal with no data in signal pending structures (no handler,
no sign of proper signal delivered). Suspend/freezer use this to
freeze the tasks as well. The task gets TIF_SIGPENDING set and is
woken up (if it has been sleeping in the kernel before) or kicked by
rescheduling IPI (if it was running on other CPU). This causes the task
to go to kernel/userspace boundary where the signal would be handled and
the task would be marked as safe in terms of live patching.

There are tasks which are not affected by this technique though. The
fake signal is not sent to kthreads. They should be handled in a
different way. Also if the task is in TASK_RUNNING state but not
currently running on some CPU it doesn't get the IPI, but it would
eventually handle the signal anyway. Last, if the task runs in the kernel
(in TASK_RUNNING state) it gets the IPI, but the signal is not handled
on return from the interrupt. It would be handled on return to the
userspace in the future when the fake signal is sent again.

If the task was sleeping in a syscall it would be woken by our fake
signal, it would check if TIF_SIGPENDING is set (by calling
signal_pending() predicate) and return ERESTART* or EINTR. Syscalls with
ERESTART* return values are restarted in case of the fake signal (see
do_signal()). EINTR is propagated back to the userspace program. This
could disturb the program, but...

  * each process dealing with signals should react accordingly to EINTR
    return values.
  * syscalls returning EINTR happen to be quite common situation in the
    system even if no fake signal is sent.
  * freezer sends the fake signal and does not deal with EINTR anyhow.
    Thus EINTR values are returned when the system is resumed.

The very safe marking is done in entry.S on syscall and interrupt/exception
exit paths.

Note that the fake signal is not sent to stopped/traced tasks. Such task
prevents the patching to finish till it continues again (is not traced
anymore).

mb: wake up kthreads periodically {
	We call wake_up_process() for kthreads only once during the
	patching (or after it to be precise). Unfortunately this is
	not sufficient. A kthread is marked as non-migrated (its
	TIF_KGR_IN_PROGRESS is set) and wake_up_process() is called on
	this kthread after the patching. However if the kthread is
	running, wake_up_process() has no effect and the kthread can
	go to sleep without a call to kgr_task_safe(). Thus the
	kthread would block the patching process from finishing till
	it is woken up again by some other process than kgraft. We can
	fix this by calling wake_up_process() periodically in the same
	way we do it with sending a fake signal.
}

Signed-off-by: Miroslav Benes <mbenes@suse.cz>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 kernel/kgraft.c |   35 +++++++++++++++++++++++++++++++++++
 kernel/signal.c |    3 ++-
 2 files changed, 37 insertions(+), 1 deletion(-)

--- a/kernel/kgraft.c
+++ b/kernel/kgraft.c
@@ -270,6 +270,39 @@ static void kgr_finalize(void)
 	mutex_unlock(&kgr_in_progress_lock);
 }
 
+static void kgr_send_fake_signal(void)
+{
+	struct task_struct *p, *t;
+
+	read_lock(&tasklist_lock);
+	for_each_process_thread(p, t) {
+		if (!klp_kgraft_task_in_progress(t))
+			continue;
+
+		/*
+		 * There is a small race here. We could see TIF_KGR_IN_PROGRESS
+		 * set and decide to wake up a kthread or send a fake signal.
+		 * Meanwhile the thread could migrate itself and the action
+		 * would be meaningless.  It is not serious though.
+		 */
+		if (t->flags & PF_KTHREAD) {
+			/*
+			 * Wake up a kthread which still has not been migrated.
+			 */
+			wake_up_process(t);
+		} else {
+			/*
+			 * Send fake signal to all non-kthread tasks which are
+			 * still not migrated.
+			 */
+			spin_lock_irq(&t->sighand->siglock);
+			signal_wake_up(t, 0);
+			spin_unlock_irq(&t->sighand->siglock);
+		}
+	}
+	read_unlock(&tasklist_lock);
+}
+
 static void kgr_work_fn(struct work_struct *work)
 {
 	static bool printed = false;
@@ -281,6 +314,8 @@ static void kgr_work_fn(struct work_stru
 				KGR_TIMEOUT);
 			printed = true;
 		}
+		/* send fake signal */
+		kgr_send_fake_signal();
 		/* recheck again later */
 		queue_delayed_work(kgr_wq, &kgr_work, KGR_TIMEOUT * HZ);
 		return;
--- a/kernel/signal.c
+++ b/kernel/signal.c
@@ -157,7 +157,8 @@ void recalc_sigpending_and_wake(struct t
 
 void recalc_sigpending(void)
 {
-	if (!recalc_sigpending_tsk(current) && !freezing(current))
+	if (!recalc_sigpending_tsk(current) && !freezing(current) &&
+	    !klp_kgraft_task_in_progress(current))
 		clear_thread_flag(TIF_SIGPENDING);
 
 }
